### PR TITLE
Update register_queue.py to enable registration of group names with unicode

### PR DESCRIPTION
**Pull request by beeankha**
##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

 - API
awx/main/management/commands/register_queue.py

##### AWX VERSION

```
AWX 2.1.0
```

##### ADDITIONAL INFORMATION

There was a bug that prevented registrations of instance groups that contained unicode.

```
failed: [localhost] (item=[] + [ 'fröbözz' ] + [ 'xyz' ] + [ '🙈🙉🙊' ] + [ 'tower' ]) => {"changed": false,
 "cmd": "awx-manage register_queue --queuename=\"[] + [ 'fröbözz' ] + [ 'xyz' ] + [ '🙈🙉🙊' ] + [ 'tower' ]\" --hostnames=\"localhost\"",
 "delta": "0:00:04.766854",
 "end": "2018-01-23 20:57:58.086117",
 "item": "[] + [ 'fröbözz' ] + [ 'xyz' ] + [ '🙈🙉🙊' ] + [ 'tower' ]",
 "msg": "non-zero return code",
 "rc": 1,
 "start": "2018-01-23 20:57:53.319263",
 "stderr": "Traceback (most recent call last):
  File \"/bin/awx-manage\", line 9, in <module>
    load_entry_point('awx==3.2.2', 'console_scripts', 'awx-manage')()
  File \"/lib/python2.7/site-packages/awx/__init__.py\", line 107, in manage
  File \"/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/management/__init__.py\", line 354, in execute_from_command_line
    utility.execute()
  File \"/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/management/__init__.py\", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File \"/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/management/base.py\", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File \"/var/lib/awx/venv/awx/lib/python2.7/site-packages/django/core/management/base.py\", line 445, in execute
    output = self.handle(*args, **options)
  File \"/lib/python2.7/site-packages/awx/main/management/commands/register_queue.py\", line 28, in handle
  File \"/usr/lib64/python2.7/contextlib.py\", line 17, in __enter__
    return self.gen.next()
  File \"/lib/python2.7/site-packages/awx/main/utils/pglock.py\", line 13, in advisory_lock
  File \"/usr/lib64/python2.7/contextlib.py\", line 17, in __enter__
    return self.gen.next()
  File \"/var/lib/awx/venv/awx/lib/python2.7/site-packages/django_pglocks/__init__.py\", line 47, in advisory_lock
    pos = crc32(lock_id.encode(\"utf-8\"))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 38: ordinal not in range(128)", 
...
```

Showing output error + correct output after change:

outputfixedafterchange

Instances successfully registered:

instancesapi

### CODE REVIEW

File: awx/main/management/commands/register_queue.py [hunk 1]
```diff
@@ -104,7 +104,7 @@ def handle(self, **options):
             changed3 = False
             (ig, created, changed1) = self.get_create_update_instance_group(queuename, inst_per, inst_min)
             if created:
-                print(six.text_type("Creating instance group {}".format(ig.name)))
+                print(six.text_type("Creating instance group {}").format(ig.name))
```
Comment (ryanpetrello) [inline awx/main/management/commands/register_queue.py:107]: D'oh!  Pretty sure this one's my fault!

**Comment by wenottingham** [comment]
To fully fix this, we'll need to also do something with `awx/main/modles/ha.py:validate_queuename()`, which prevents editing IGs with unicode names.

**Comment by ryanpetrello** [comment]
If this actually works now, we can probably lose this code:

https://github.com/ansible/awx/blob/devel/awx/main/models/ha.py#L34-L40
https://github.com/ansible/awx/blob/devel/awx/main/models/ha.py#L166-L168
https://github.com/ansible/awx/blob/devel/awx/main/models/ha.py#L237-L239

Before merging this, I'd also like to update integration to actually run an install w/ a few instances/groups that contain unicode names cc @jladdjr

**Comment by ryanpetrello** [comment]
As @wenottingham alluded to, it would also be nice to write some tests that actually hit the API endpoints and try to _set_ instance and group names that contain unicode (and verified that we got an HTTP 200/201 back).

https://github.com/ansible/awx/blob/devel/awx/main/tests/functional/api/test_instance_group.py is probably a good place to start looking.

**Comment by jladdjr** [comment]
I'm not sure I have a reference to a ticket mentioning this, but I vaguely remember this coming up before and the team deciding that unicode names just weren't allowed with instance groups. Maybe that's what @wenottingham was talking about in his comment (`To fully fix this, we'll need to also do something with awx/main/modles/ha.py:validate_queuename(), which prevents editing IGs with unicode names.`) Does anyone else remember us having a conversation like that? If it doesn't matter any more and we just need to run tests, that's fine, I just thought this was something that had already come up.

**Comment by ryanpetrello** [comment]
@jladdjr

> Does anyone else remember us having a conversation like that?

Yep, this all ended up happening informally at some point last year - we realized that celery and kombu wouldn't let us do this, so we bailed.  Now that we're on newer versions of this library, it looks like this PR (and the removal of the validation code in `ha.py`) is what we need to allow unicode in group names.

**Comment by jladdjr** [comment]
@ryanpetrello (cc @beeankha) - roger that, shouldn't be hard to update our tests to exercise unicode. Think it might just involve a small change in towerkit. Checking that out now.

**Comment by jladdjr** [comment]
>  it would also be nice to write some tests that actually hit the API endpoints and try to set instance and group names

@ryanpetrello - silly question, what would it look like to set an _instances_ name using unicode? afaict, instances have hostnames, but otherwise aren't actually named:
```
[{
     "jobs_running": 0,
     "consumed_capacity": 0,
     "cpu_capacity": 8,
     "jobs_total": 37,
     "related": {
         "instance_groups": "/api/v2/instances/1/instance_groups/",
         "jobs": "/api/v2/instances/1/jobs/"
     },
     "percent_capacity_remaining": 100.0,
     "id": 1,
     "mem_capacity": 54,
     "capacity_adjustment": "1.00",
     "capacity": 54,
     "uuid": "62b8f876-fcf4-42c1-9eff-f46cbc84923f",
     "created": "2018-11-05T19:32:18.627613Z",
     "url": "/api/v2/instances/1/",
     "hostname": "localhost",   <-----
     "enabled": true,
     "modified": "2018-11-14T18:16:32.296774Z",
     "managed_by_policy": true,
     "version": "3.4.0",
     "memory": 7891697664,
     "type": "instance",
     "cpu": 2
 }]
```

**Comment by ryanpetrello** [comment]
@jladdjr that's a great question, and I don't know the answer 😂 

If we could get unicode support for Instance *and* Instance Groups out of this PR, that would be fantastic.  Either way, I'd like to have some integration tests that confirm behavior so we know where we stand and can make an estimation of what's left/missing.

**Comment by jladdjr** [comment]
* From what I can tell, this change only affects the handling of queues, not seeing anything related to instances
* Based on [1] and [2], it sounds like even when FQDNs make use of internationalization, ASCII is still used under the hood using a hacky conversion scheme (see Punycode)

---

[1] https://en.wikipedia.org/wiki/Internationalized_domain_name
[2] https://en.wikipedia.org/wiki/Punycode

**Comment by ryanpetrello** [comment]
@jladdjr nice!

**Comment by jladdjr** [comment]
Lo and behold, a cluster install just passed (cc @shanemcd). So, if we can just get the `ha.py` change that @wenottingham mentioned in, we could try exercising the change.

**Comment by beeankha** [comment]
I'll submit that `ha.py` change shortly!

**Comment by jladdjr** [comment]
Will create instance group with unicode name when we deploy tower:
https://github.com/ansible/tower-qa/pull/2483

**Comment by ryanpetrello** [comment]
Hey @beeankha,

This is almost ready to go.  CI is complaining about some lint/style errors; it looks like there's some Python imports you can remove (since they're no longer being used):

```
2018-11-14 22:44:13.979631 | static | ./awx/main/models/ha.py:4:1: F401 'six' imported but unused
2018-11-14 22:44:13.979778 | static | ./awx/main/models/ha.py:8:1: F401 'django.core.exceptions.ValidationError' imported but unused
```

If you'll address these, I'll 👍 this PR.

**Comment by ryanpetrello** [comment]
This might help things pass: https://github.com/ryanpetrello/awx/commit/4c4ba334a209875c69710db9c9c6095008c590bf

Testing w/ @jladdjr...

**Comment by elyezer** [comment]
@ryanpetrello after https://github.com/ansible/awx/pull/2538 being merged I think this will need some update, right?

Also I will be following this up. I got this from @jladdjr

**Comment by ryanpetrello** [comment]
@elyezer @jladdjr @beeankha Yep, we're holding off on this until py3 merges.  Once that PR merges, a lot of the code here will be unnecessary.

**Comment by ryanpetrello** [comment]
Now that py3 support merged, this is ready to be rebased and looked at again.

**Comment by ryanpetrello** [comment]
@beeankha I'm gonna close this since you've opened https://github.com/ansible/awx/pull/3020 as a replacement.
